Unify toploop.mli / opttoploop.mli

AltGr opened the pull request:
This unifies both implementations of the toplevel under a common interface.

Similar Makefile hacks to those in place for dynlink are used: byte
and native implementations are in separate directories, while the
shared interface are in the parent and copied as needed by the build
rules.

This is a preliminary step to remove duplication as much as possible
in the two implementations, but should already be pretty useful for
libraries that use the `Toploop` interface.

AltGr committed:
Unify toploop.mli / opttoploop.mli

This unifies both implementations of the toplevel under a common interface.

Similar Makefile hacks to those in place for dynlink are used: byte
and native implementations are in separate directories, while the
shared interface are in the parent and copied as needed by the build
rules.

This is a preliminary step to remove duplication as much as possible
in the two implementations, but should already be pretty useful for
libraries that use the `Toploop` interface.

gasche commented:
> Similar Makefile hacks to those in place for dynlink are used: byte
and native implementations are in separate directories, while the
shared interface are in the parent and copied as needed by the build
rules.

Naive question: could we define the interface in a `Toplevel_sigs` module (that would define one signature for each of the .mli we want to share), and then simply have `.mli` files of the form `include Toplevel_sigs.Foo`?

AltGr committed:
Remove stale file

ghost commented:
That would make sense if we planned to keep the opttop* modules. However, the end goal is to have a single set of modules and get rid of the opttop* ones.

jeremiedimino commented on toplevel/byte/topdirs.mli:
```diff
@@ -26,6 +26,9 @@ val dir_use : formatter -> string -> unit
 val dir_use_output : formatter -> string -> unit
```
Looks like this file is the same as `toplevel/topdirs.mli`. I guess it should be deleted?

AltGr replied:
Ah yes, that was a stray file, did'nt I push that fix already ?
Thanks

jeremiedimino replied:
Just updated and it's gone now

gasche commented:
> That would make sense if we planned to keep the opttop* modules. However, the end goal is to have a single set of modules and get rid of the opttop* ones.

So, if I understand correctly, the Makefile blurb is here only for a transitory period, and will be un-complexified after the code has been factorized appropriately? The plan is that, in the end, there is a single module that uses a dynamic check/option to behave as either the bytecode or native toplevel?

AltGr commented:
Yes, this is a first stage in merging as much as possible from the `toplevel/byte` and `toplevel/native` directories ;
I'm not happier with the Makefile blurbs than you seem to be :nauseated_face: (actually I don't think I would have dared do anything like that if it was not for the "nice" example in `compilerlibs/dynlink` :roll_eyes: ).

On the outside, having a unified interface should already be an improvement though. And quite a few upgrades of the bytecode toplevel are being ported to the native one as it goes, so that's a benefit as well.

AltGr committed:
Fix test reference

the source file location path changed...

ghost commented:
> The plan is that, in the end, there is a single module that uses a dynamic check/option to behave as either the bytecode or native toplevel?

Either dynamic tests, either a module with a different implementation in byte code and native. The first option is simpler to implement, but it means that you'd end up linking the native compilation backend in byte code programs using the toplevel API, which is slightly odd.

ghost commented:
One question regarding the `Trace` module;  IIUC we continue to have it only in bytecode. That means that we cannot write a program that refers to `Trace` and have this program build out of the box in native. The programmer would need to resort to optional compilation or virtual libraries or something of the sort to be able to make their program work in both byte code and native code if they want to use `Trace` in bytecode.

Another option would be to systematically have a `Trace` module, but have all the functions fail with `failwith "Not implemented"` in native. IIRC, this is similar with what we do for some unix functions on Windows. WDYT?

jeremiedimino commented on compilerlibs/Makefile.compilerlibs:
```diff
@@ -261,14 +261,32 @@ MIDDLE_END_CMI=\
+TOPLEVEL_SHARED_MLIS = toploop.mli topdirs.mli topmain.mli
+
+toplevel/%/toploop.cmi: toplevel/toploop.cmi
+	cp $< toplevel/toploop.mli $(@D)
```
I can believe that this works, but it seems unnecessarily fishy to me. i.e. we don't declare one of the target.

Do you mind splitting this rule (as well as the next two) in two simpler copy rules?

AltGr replied:
Indeed, it should work because the `cmi` is necessary more recent than the `mli` at this point, but...

The same is done [here](https://github.com/ocaml/ocaml/blob/7bbde5be870cdbd20281e3b7f0fcb2b36f755373/otherlibs/dynlink/Makefile#L199) ; but I'll test and change if there are no issues to two separate rules.

AltGr replied:
Please tell me what you think of the new version; it's clearer and cleaner, but still slightly suspicious because it theoretically allows copying any mli/cmi from `toplevel/` to `toplevel/{byte,native}/` ; in practive it won't because of the rules generated by `make depends`, which we ensure are correct with the explicit `beforedepends` copy rule.

jeremiedimino replied:
We discussed this today with Louis. I was expecting that splitting the rule in two would be better, but it turns out that it leads to inconsistent assumptions. Given that the way the rules are currently written works and we do that in other parts of the Makefiles, we decided to leave them as is.

AltGr commented:
Indeed, in the current state of the PR, the `Trace` module only exists when compiling to bytecode.

> Another option would be to systematically have a `Trace` module, but have all the functions fail with `failwith "Not implemented"` in native. IIRC, this is similar with what we do for some unix functions on Windows. WDYT?

This is what I have been doing in the current version for the `Topdirs.dir_*trace*` [functions](https://github.com/ocaml/ocaml/pull/10061/files#diff-c92c3841c65d65721068086e1cd8ed662c29f5428643296e082eded844f4059cR200-R202), but it's true that I didn't consider uses of the `Trace` module directly. I can make it shared as well, with a dummy implementation for native ; it may depend on the consumers of this module, I don't mind either way.

As for the `topdirs`, I have a second version underway where common directives are shared and each implementation defines its own (`trace`, etc.) in addition.

AltGr committed:
Cleanup toplevel byte/native Makefile copy rules

Copying the file separately causes issues (there would be two paths to build
`toplevel/byte/toploop.cmi`); this feels a bit cleaner, but the rules are not
guarded anymore; it is ok because of the dependency rules ensuing from our
`beforedepend` rule.

ghost commented:
> This is what I have been doing in the current version for the Topdirs.dir_*trace* functions, but it's true that I didn't consider uses of the Trace module directly. I can make it shared as well, with a dummy implementation for native ; it may depend on the consumers of this module, I don't mind either way.

Ack. Let's make it shared with a dummy implementation for native; it's more uniform and it will make things easier for consumers.

> As for the topdirs, I have a second version underway where common directives are shared and each implementation defines its own (trace, etc.) in addition.

Sounds good

AltGr committed:
Provide a dummy "Trace" module in the native toplevel

ghost commented:
We just had another look with @AltGr. Just need to fix CI issues and add a disclaimer in `trace.mli` to say that this feature is not available in native mode, and then it looks good to merge.

AltGr committed:
Fix install rules, add visible disclaimer in toplevel/trace.mli

ghost commented:
We need a changelog entry BTW. There is one visible change: modules of the opttoplevel library are renamed `OptXXX -> XXX`

AltGr committed:
Update Changes

AltGr committed:
Remove stale mli file

AltGr committed:
Makefiles 80 columns

AltGr committed:
make alldepend

ghost commented:
All looking good from my point of view! @AltGr could you confirm you are happy with the current state of the PR? I'll merge if yes.

AltGr commented:
Yep! Still some work to do to clean up the code duplication, but this is fine as a first step forward :)

ghost commented:
Great, merging!
